### PR TITLE
Moves C++ Python utility code into `PythonInterfaceCore`

### LINKED ISSUE #26334: Move C++ Python utility code out of kernel to PythonInterfaceCore libary
The `_kernel` library should be purely the exports required for the module. All utility types and functions should be moved to `PythonInterfaceCore`

### PR DESCRIPTION
**Pull request by dtasev**
**Description of work.**
Moves some utility C++ Python classes into the `PythonInterfaceCore` target

**To test:**
Builds should pass. MantidPlot and Workbench should start as usual

Fixes #26334

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/core/CMakeLists.txt [hunk 1]
```diff
@@ -4,69 +4,125 @@
...
+  inc/MantidPythonInterface/core/Policies/MatrixToNumpy.h
+  inc/MantidPythonInterface/core/Policies/RemoveConst.h
+  inc/MantidPythonInterface/core/Policies/ToWeakPtr.h
+  inc/MantidPythonInterface/core/Policies/VectorToNumpy.h
+)
 
 # Add the target for this directory
 set(_target_name PythonInterfaceCore)
-add_library(${_target_name} ${SRC_FILES} ${INC_FILES})
+add_library(
```
Comment (DTasev) [inline Framework/PythonInterface/core/CMakeLists.txt]: changes here and below due to `cmake-format`

**Review by martyngigg — COMMENTED** [review]
This is good. Do you think we can move the headers for `api` and `kernel` into the respective directories and get rid of the `PythonInterface/inc` directory?

**Comment by DTasev** [comment]
Done in the latest commit, mostly renaming of files, but also removed `HEADER_DIR` variable

File: Framework/PythonInterface/mantid/geometry/CMakeLists.txt [hunk 2]
```diff
@@ -53,7 +53,9 @@ create_module(${MODULE_TEMPLATE} ${MODULE_DEFINITION} ${EXPORT_FILES})
 
 set(SRC_FILES)
 
-set(INC_FILES)
+set(INC_FILES
+  inc/MantidPythonInterface/geometry/ComponentInfoPythonIterator.h
```
Comment (DTasev) [inline Framework/PythonInterface/mantid/geometry/CMakeLists.txt]: These 2 are never used outside the `PythonInterface/geometry` module. They were included in the `api` which now failed as `api` is dependent on `geometry`, so they weren't available in the includes.

**Comment by martyngigg** [comment]
@DTasev It looks like there are some doxygen warnings.

**Review by martyngigg — DISMISSED** [review]
I'm happy with this now. It removes the `PythonInterface/inc` directory that was never really meant to be included by other projects. The `PythonInterfaceCore` library is now the place to include common Python code for the whole project.

**Comment by DTasev** [comment]
Unfortunately I can't give an ETA, but I am hoping to have a look by the end of the week.

**Comment by martyngigg** [comment]
No worries. No rush.

**Comment by DTasev** [comment]
I think that should do it - I pointed Doxygen to the new location of the includes.

**Review by martyngigg — APPROVED** [review]

**Comment by martyngigg** [comment]
@mantidproject/gatekeepers Could we get this merged, please?

**Pull request merged**